File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/common/OrganizationService.java
package com.jasify.schedule.appengine.model.common;

import com.google.appengine.api.datastore.Key;
import com.jasify.schedule.appengine.model.EntityNotFoundException;
import com.jasify.schedule.appengine.model.FieldValueException;
import com.jasify.schedule.appengine.model.UniqueConstraintException;
import com.jasify.schedule.appengine.model.users.User;

import javax.annotation.Nonnull;
import java.util.List;

/**
 * @author krico
 * @since 08/01/15.
 */
public interface OrganizationService {

    /**
     * @param group to be added
     * @return the ide of the newly created group
     * @throws UniqueConstraintException if the group name already existed withing this context (not so sure ATM)
     * @throws FieldValueException       if any fields had invalid values (e.g name missing)
     */
    @Nonnull
    public Key addGroup(Group group) throws UniqueConstraintException, FieldValueException;

    /**
     * @param id of the group
     * @return the group
     * @throws EntityNotFoundException  if the group doesn't exist
     * @throws IllegalArgumentException if <code>id</code> is not the key to a group
     */
    @Nonnull
    public Group getGroup(Key id) throws EntityNotFoundException, IllegalArgumentException;

    /**
     * @return all groups
     */
    @Nonnull
    public List<Group> getGroups();

    /**
     * @param group to update
     * @return the updated group
     * @throws EntityNotFoundException if the group didn't exist before
     * @throws FieldValueException     if fields have invalid values
     */
    @Nonnull
    public Group updateGroup(Group group) throws EntityNotFoundException, FieldValueException;

    /**
     * @param group to add the user to
     * @param user  to be added to the group
     * @throws EntityNotFoundException if either don't exist
     */
    public void addUserToGroup(Group group, User user) throws EntityNotFoundException;

    /**
     * @param groupId to add to
     * @param userId  to be added
     * @throws EntityNotFoundException  if either doesn't exist
     * @throws IllegalArgumentException if either is not the entity they are expected to be
     */
    public void addUserToGroup(Key groupId, Key userId) throws EntityNotFoundException, IllegalArgumentException;

    /**
     * @param group to remove from
     * @param user  to add
     * @throws EntityNotFoundException if either doesn't exist
     */
    public void removeUserFromGroup(Group group, User user) throws EntityNotFoundException;

    /**
     * @param groupId to remove from
     * @param userId  to remove
     * @throws EntityNotFoundException  if any doesn't exist
     * @throws IllegalArgumentException if any is not the expected entity
     */
    public void removeUserFromGroup(Key groupId, Key userId) throws EntityNotFoundException, IllegalArgumentException;

    /**
     * Remove and existing organization
     *
     * @param id of the group to be removed
     * @throws EntityNotFoundException  if the group doesn't exist
     * @throws IllegalArgumentException if <code>id</code> is not the key to a Group
     */
    public void removeGroup(Key id) throws EntityNotFoundException, IllegalArgumentException;
}

File: schedule/schedule-appengine/src/main/client/jasify/services/organization.service.js
(function (angular) {
    angular.module('jasifyComponents').factory('Organization', organization);

    function organization(Endpoint, $q, $log) {
        var Organization = {
            query: query,
            queryPublic: queryPublic,
            get: get,
            update: update,
            add: add,
            users: users,
            groups: groups,
            addUser: addUser,
            removeUser: removeUser,
            addGroup: addGroup,
            removeGroup: removeGroup,
            remove: remove
        };

        function query() {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.query()
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function queryPublic() {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.queryPublic()
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function get(id) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.get({id: id})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function update(organization) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.update(organization)
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function addUser(organization, user) {
            var organizationId = angular.isObject(organization) ? organization.id : organization;
            var userId = angular.isObject(user) ? user.id : user;

            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.addUser({organizationId: organizationId, userId: userId})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function removeUser(organization, user) {
            var organizationId = angular.isObject(organization) ? organization.id : organization;
            var userIds = [];
            if (!angular.isArray(user)) user = [user];

            for (var x = 0; x < user.length; ++x) {
                var userId = angular.isObject(user[x]) ? user[x].id : user[x];
                userIds.push(userId);
            }

            var promises = [];
            for (var i = 0; i < userIds.length; ++i) {
                promises.push(removeOneUser(organizationId, userIds[i]));
            }

            function removeOneUser(organizationId, userId) {
                return Endpoint.jasify(function (jasify) {
                    return jasify.organizations.removeUser({organizationId: organizationId, userId: userId})
                        .then(Endpoint.resultHandler, Endpoint.rejectHandler);
                });
            }

            if (promises.length == 1) return promises[0];
            return $q.all(promises);
        }

        function addGroup(organization, group) {
            var organizationId = angular.isObject(organization) ? organization.id : organization;
            var groupId = angular.isObject(group) ? group.id : group;

            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.addGroup({organizationId: organizationId, groupId: groupId})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function removeGroup(organization, group) {
            var organizationId = angular.isObject(organization) ? organization.id : organization;
            var groupIds = [];
            if (!angular.isArray(group)) group = [group];

            for (var x = 0; x < group.length; ++x) {
                var groupId = angular.isObject(group[x]) ? group[x].id : group[x];
                groupIds.push(groupId);
            }

            var promises = [];
            for (var i = 0; i < groupIds.length; ++i) {
                promises.push(removeOneGroup(organizationId, groupIds[i]));
            }

            function removeOneGroup(organizationId, groupId) {
                return Endpoint.jasify(function (jasify) {
                    return jasify.organizations.removeGroup({organizationId: organizationId, groupId: groupId})
                        .then(Endpoint.resultHandler, Endpoint.rejectHandler);
                });
            }

            if (promises.length == 1) return promises[0];
            return $q.all(promises);
        }

        function add(organization) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.add(organization)
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function users(id) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.users({id: id})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function groups(id) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.groups({id: id})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        function remove(id) {
            return Endpoint.jasify(function (jasify) {
                return jasify.organizations.remove({id: id})
                    .then(Endpoint.resultHandler, Endpoint.rejectHandler);
            });
        }

        return Organization;
    }

})(angular);